### ISSUE TITLE
Confusing dialog for contributed launch configs when errors occur

### ISSUE THREAD

**Issue by sbatten** [microsoft/vscode#156351]
Testing https://github.com/microsoft/vscode-python/issues/19547

Steps:
1. Create a basic python script file in a new folder
2. Have the python extension installed
3. Go to the debug view container and select the "Show all automatic configurations" option
4. Choose Python... and File
5. Open the settings editor
6. Run the dynamic launch config

🐛 get a dialog about the ${file} variable and a button to open the launch.json

Neither of these apply since you didn't create the launch configuration nor does it exist in the launch.json

**Comment by roblourens** [comment]
I think that the message about "open an editor" will make sense to people, they should know what we mean. I can hide the launch.json button though.

**Comment by sbatten** [comment]
@roblourens In this case, the user didn't create any configuration. They got this auto configuration when they had the python file open. So they aren't necessarily familiar with the ${file} variable nor what an "editor" is. VS Code is open, so what does open an editor mean? Can we be more explicit? e.g. "A file must be open and ~active for running this configuration. Please open a file and try again".

**Comment by roblourens** [comment]
@alexr00 would it make sense to give the variable resolver a mode where it throws alternate error messages that don't assume that variable came from something that the user has control over?

I'm not sure, I think people will still figure out what this means

**Comment by roblourens** [comment]
Or you could put an id/enum field on the errors you throw so that the debug service can swap in its own error message.

**Comment by alexr00** [comment]
> @alexr00 would it make sense to give the variable resolver a mode where it throws alternate error messages that don't assume that variable came from something that the user has control over?

The variable resolver only knows about variables. I think we always include the variable that failed in the error. Can you check for the specific variable that failed and show your own message based on that?

**Comment by roblourens** [comment]
You mean just check whether it shows up in the error message? I don't want to rely on that. I'm looking at these errors:

https://github.com/microsoft/vscode/blob/cfd3c5c86e14552cc01bf926b09395dcfe26ce4d/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L199

**Comment by alexr00** [comment]
Yes, that's the error I was referring too. If the `match` from that line is `${file}` then you could use your own error message instead of the one from the variable resolver. We can also expand upon the `Please open an editor` if that would help.

**Comment by roblourens** [comment]
`match` is just getting plugged into the localized error message. I don't think it's safe to rely on a substring in an error message

**Comment by alexr00** [comment]
@roblourens what do you think about something like this: https://github.com/microsoft/vscode/pull/157788

If that works for you then I'll add error kinds for other variables, not just `file`.

**Comment by roblourens** [comment]
Reopened for me to adopt

**Comment by roblourens** [comment]
Is there anything similar in tasks or any other consumer? A case where an extension generates a config with variables that get resolved, and the user never sees the variables?

**Comment by alexr00** [comment]
@meganrogge for tasks. 

Yes, an extension that provides tasks can include variables in their task. If the user hasn't configured the tasks in tasks.json then you could end up in this situation.

**Comment by roblourens** [comment]
I guess the error message still has to be specific to debugging, which is what the user knows about, and we probably can't provide a useful generic error message that doesn't include the specific variable.

**Comment by roblourens** [comment]
Oh no. Sorry for taking so long to try to adopt this but I just learned something - you can't throw custom error types across our IPC. The variable resolver is running in the EH and in the renderer process, any extra properties on the error have just been lost.

**Comment by roblourens** [comment]
Well I should be able to capture the custom error type and send it back as a value, this all goes through the ExtHostDebugService. Should be no problem.
